File: split.js
const fs = require('fs');
const path = require('path');

const PAGE_SIZE = 10;
const inputFile = path.join(__dirname, 'wenzhang.json');
const outputDir = path.join(__dirname, 'articles');

if (!fs.existsSync(outputDir)) {
    fs.mkdirSync(outputDir, { recursive: true });
}

let raw;
try {
    raw = fs.readFileSync(inputFile, 'utf-8');
} catch {
    console.log('未找到 wenzhang.json，跳过拆分');
    process.exit(0);
}

let data;
try {
    data = JSON.parse(raw);
} catch {
    console.error('wenzhang.json 格式错误');
    process.exit(1);
}

// 读取原始数组，并给每个文章打上原始索引作为 id
let rawArticles = (data.announcements || []).map((article, index) => {
    return { ...article, id: index };
});

// 过滤 delete: true 的文章
let articles = rawArticles.filter(a => a.delete !== true);

if (articles.length === 0) {
    console.log('没有文章，跳过拆分');
    process.exit(0);
}

// 按日期升序排序（旧→新）
const sorted = articles.slice().sort((a, b) => {
    const da = a.date || '1970-01-01';
    const db = b.date || '1970-01-01';
    return da.localeCompare(db);
});

const total = sorted.length;
const totalPages = Math.ceil(total / PAGE_SIZE);

// 生成分页文件
for (let i = 0; i < totalPages; i++) {
    const start = i * PAGE_SIZE;
    const end = Math.min(start + PAGE_SIZE, total);
    const pageData = {
        total,
        totalPages,
        page: i + 1,
        list: sorted.slice(start, end)
    };
    fs.writeFileSync(
        path.join(outputDir, `page-${i + 1}.json`),
        JSON.stringify(pageData, null, 2)
    );
    console.log(`生成 page-${i + 1}.json (${end - start} 篇)`);
}

// 生成 all.json（完整数据，包含 id）
fs.writeFileSync(
    path.join(outputDir, 'all.json'),
    JSON.stringify({ total, list: sorted }, null, 2)
);
console.log(`生成 all.json (${total} 篇)`);

// 生成 archive.json（轻量数据，不含正文，但保留 id）
const archiveList = sorted.map(a => ({
    id: a.id,
    title: a.title || '无标题',
    date: a.date || '1970-01-01',
    tags: a.tags || []
}));
fs.writeFileSync(
    path.join(outputDir, 'archive.json'),
    JSON.stringify({ total, list: archiveList }, null, 2)
);
console.log(`生成 archive.json (${total} 篇，无正文)`);

// 生成单篇文章文件，文件名使用原始 id
sorted.forEach(article => {
    fs.writeFileSync(
        path.join(outputDir, `article-${article.id}.json`),
        JSON.stringify(article, null, 2)
    );
});
console.log(`生成 ${total} 个独立文章文件`);

// 元数据
fs.writeFileSync(
    path.join(outputDir, 'meta.json'),
    JSON.stringify({ total, totalPages }, null, 2)
);
console.log(`生成 meta.json (总页数 ${totalPages})`);

console.log(`拆分完成：${total} 篇文章，${totalPages} 页`);

File: split.go
// +build ignore

package main

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
)

const PAGE_SIZE = 10

func main() {
	inputFile := "wenzhang.json"
	outputDir := "articles"

	if err := os.MkdirAll(outputDir, 0755); err != nil {
		fmt.Fprintf(os.Stderr, "创建目录失败: %v\n", err)
		os.Exit(1)
	}

	raw, err := os.ReadFile(inputFile)
	if err != nil {
		if os.IsNotExist(err) {
			fmt.Println("未找到 wenzhang.json，跳过拆分")
			os.Exit(0)
		}
		fmt.Fprintf(os.Stderr, "读取文件失败: %v\n", err)
		os.Exit(1)
	}

	var data map[string]interface{}
	if err := json.Unmarshal(raw, &data); err != nil {
		fmt.Fprintf(os.Stderr, "wenzhang.json 格式错误\n")
		os.Exit(1)
	}

	announcements, ok := data["announcements"].([]interface{})
	if !ok {
		announcements = []interface{}{}
	}

	var rawArticles []map[string]interface{}
	for idx, item := range announcements {
		article, ok := item.(map[string]interface{})
		if !ok {
			continue
		}
		newArticle := make(map[string]interface{})
		for k, v := range article {
			newArticle[k] = v
		}
		newArticle["id"] = idx
		rawArticles = append(rawArticles, newArticle)
	}

	var articles []map[string]interface{}
	for _, a := range rawArticles {
		if del, ok := a["delete"]; ok {
			if delBool, ok := del.(bool); ok && delBool {
				continue
			}
		}
		articles = append(articles, a)
	}

	if len(articles) == 0 {
		fmt.Println("没有文章，跳过拆分")
		os.Exit(0)
	}

	sorted := make([]map[string]interface{}, len(articles))
	copy(sorted, articles)
	sort.Slice(sorted, func(i, j int) bool {
		dateI := getStringField(sorted[i], "date", "1970-01-01")
		dateJ := getStringField(sorted[j], "date", "1970-01-01")
		return dateI < dateJ
	})

	total := len(sorted)
	totalPages := (total + PAGE_SIZE - 1) / PAGE_SIZE

	for p := 1; p <= totalPages; p++ {
		start := (p - 1) * PAGE_SIZE
		end := start + PAGE_SIZE
		if end > total {
			end = total
		}
		pageData := map[string]interface{}{
			"total":      total,
			"totalPages": totalPages,
			"page":       p,
			"list":       sorted[start:end],
		}
		filename := filepath.Join(outputDir, fmt.Sprintf("page-%d.json", p))
		if err := writeJSON(filename, pageData); err != nil {
			fmt.Fprintf(os.Stderr, "写入 %s 失败: %v\n", filename, err)
			os.Exit(1)
		}
		fmt.Printf("生成 page-%d.json (%d 篇)\n", p, end-start)
	}

	allData := map[string]interface{}{
		"total": total,
		"list":  sorted,
	}
	if err := writeJSON(filepath.Join(outputDir, "all.json"), allData); err != nil {
		fmt.Fprintf(os.Stderr, "写入 all.json 失败: %v\n", err)
		os.Exit(1)
	}
	fmt.Printf("生成 all.json (%d 篇)\n", total)

	archiveList := make([]map[string]interface{}, 0, total)
	for _, a := range sorted {
		title := getStringField(a, "title", "无标题")
		date := getStringField(a, "date", "1970-01-01")
		tags := getTagsField(a)
		archiveList = append(archiveList, map[string]interface{}{
			"id":    a["id"],
			"title": title,
			"date":  date,
			"tags":  tags,
		})
	}
	archiveData := map[string]interface{}{
		"total": total,
		"list":  archiveList,
	}
	if err := writeJSON(filepath.Join(outputDir, "archive.json"), archiveData); err != nil {
		fmt.Fprintf(os.Stderr, "写入 archive.json 失败: %v\n", err)
		os.Exit(1)
	}
	fmt.Printf("生成 archive.json (%d 篇，无正文)\n", total)

	for _, article := range sorted {
		idVal, ok := article["id"]
		if !ok {
			continue
		}
		filename := filepath.Join(outputDir, fmt.Sprintf("article-%v.json", idVal))
		if err := writeJSON(filename, article); err != nil {
			fmt.Fprintf(os.Stderr, "写入 %s 失败: %v\n", filename, err)
			os.Exit(1)
		}
	}
	fmt.Printf("生成 %d 个独立文章文件\n", total)

	metaData := map[string]interface{}{
		"total":      total,
		"totalPages": totalPages,
	}
	if err := writeJSON(filepath.Join(outputDir, "meta.json"), metaData); err != nil {
		fmt.Fprintf(os.Stderr, "写入 meta.json 失败: %v\n", err)
		os.Exit(1)
	}
	fmt.Printf("生成 meta.json (总页数 %d)\n", totalPages)

	fmt.Printf("拆分完成：%d 篇文章，%d 页\n", total, totalPages)
}

func getStringField(m map[string]interface{}, key, defaultVal string) string {
	if val, ok := m[key]; ok {
		if str, ok := val.(string); ok && str != "" {
			return str
		}
	}
	return defaultVal
}

func getTagsField(m map[string]interface{}) []interface{} {
	if val, ok := m["tags"]; ok {
		if tags, ok := val.([]interface{}); ok {
			return tags
		}
	}
	return []interface{}{}
}

func writeJSON(filename string, data interface{}) error {
	jsonData, err := json.MarshalIndent(data, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(filename, jsonData, 0644)
}
